Improve docs in relation to PhpStan failure

tomasfejfar opened the pull request:
Fixes #8

tomasfejfar committed:
Fix reported phpcs/phpstan issues in example files

tomasfejfar committed:
Run ci on example files as well

tomasfejfar committed:
Add warning that PhpStan will fail with custom config

tomasfejfar committed:
Consistent representation of example classes in README.md

tomasfejfar commented:
What do you think, @jakubbartel?

jakubbartel commented:
Great!

Just for note: when I use php-component as a composer dependency I need to use only the first of `ignoreErrors` in my project's `phpstan.neon`, but I hope this will be clear for developers (not to use all of ignored errors from php-component project).

odinuv commented on example/MyConfigDefinition.php:
```diff
@@ -1,7 +1,15 @@
 {
+    /**
+     * @return ArrayNodeDefinition|NodeDefinition
+     */
```
do we need ArrayNodeDefinition here? parameters is also always an object (or should be normalized to), so `NodeDefinition` should be enough

tomasfejfar replied:
Now that I think about it... is should be always `ArrayNodeDefinition`.  Because the parameters are always array. If it were `NodeDefinition`, it would fail because `children()` is does not exist there :(

odinuv replied:
ok, i always mix the classes, but the point is that only one of them is sufficient

ondrejhlavacek commented:
I also came across this - https://github.com/keboola/php-component/pull/10#issuecomment-370409670

It's easy to fix, no rocket science, but blind copy & paste of `phpstan.neon` won't work.

tomasfejfar commented:
I changed it little bit more. As @odinuv noted, the issue was partly with intersection typehint (which was a reminiscence from earlier version). When I removed it the PhpStan errors are gone. So I replaced the note with a [more general note](https://github.com/keboola/php-component/tree/improve-docs#custom-parameters-validation).

tomasfejfar commented:
Ještě zvažuju, jak to udělat s tagnutím nové verze. Tím, že se to srovnalo a přibyly typehinty, tak to není zpětně kompatibilní. Ale nevím, jestli kvuli tomu hned vydávat novou major verzi. Zatím to používáme stejně jen my a @jakubbartel ne? Ale formálně správně by bylo asi 2.0.0. Co myslíte?

ondrejhlavacek commented:
Klidně novou major verzi.

tomasfejfar commented:
OK, [done](https://packagist.org/packages/keboola/php-component#2.0.0).
